File: samples/FsBulletML.Sample/ParticleEmitter.fs
namespace FsBulletML.Sample

open System
open System.Collections.Generic 
open Microsoft.Xna.Framework
open Microsoft.Xna.Framework.Graphics

type ParticleEmitter () as this =
  [<DefaultValue>]val mutable private particles : List<Particle>
  [<DefaultValue>]val mutable private rand : Random

  do
    this.particles <- new List<Particle>()
    this.rand <- new Random()

  member this.Clear () =
    this.particles.Clear()

  member this.Emmit (texture, pos, scale, shrinkRate, duration, amount, maxSpeed, colour) =
    for i in [0..amount-1]  do
      let angle = this.rand.Next(0, 360) |> float32
      let speed = this.rand.Next(1, maxSpeed) |> float32
      let p = new Particle(texture, pos, speed, angle, scale, shrinkRate, duration, colour)
      if this.particles |> Seq.length < 1000 then
        this.particles.Add(p)

  member this.Update(delta:float) =
    let toRemove = new List<Particle>()
    this.particles |> Seq.iter (fun p -> p.Update(delta); if p.alive |> not then toRemove.Add(p))
    for i in [0..toRemove.Count-1] do
      this.particles.Remove(toRemove.[i]) |> ignore

  member this.Draw(sp:SpriteBatch) =
    this.particles |> Seq.iter (fun p -> p.Draw(sp))

File: samples/FsBulletML.Sample.MonoGame.CSharp/ParticleEmitter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace FsBulletML.Sample.MonoGame.CSharp
{
    class ParticleEmitter
    {
        private List<Particle> particles;
        private static Random Rand = new Random();
 
        public ParticleEmitter()
        {
            this.particles = new List<Particle>();
        }
        
        public void Emmit(Texture2D tex, Vector2 pos, float scale, float shrinkRate,float duration, int amount, int maxSpeed, Color colour)
        {
            for (int i = 0; i < amount; i++)
            {
                int angle = ParticleEmitter.Rand.Next(0, 360);
                float speed = ParticleEmitter.Rand.Next(1, maxSpeed);
                var p = new Particle(tex, pos, speed, angle, scale, shrinkRate, duration, colour);
                this.particles.Add(p);
            }
        }

        public void Clear()
        {
            this.particles.Clear();
        }

        public void Update(double delta)
        {
            List<Particle> toRemove = new List<Particle>();
            for (int i = 0; i < this.particles.Count; i++)
            {
                this.particles[i].Update(delta);
 
                if (!this.particles[i].IsActive)
                    toRemove.Add(this.particles[i]);
            }
 
            for (int i = 0; i < toRemove.Count; i++)
                this.particles.Remove(toRemove[i]);
        }

        public void Draw(SpriteBatch sp)
        {
            for (int i = 0; i < this.particles.Count; i++)
                this.particles[i].Draw(sp);
        }
    }
}
